msg/templates/uorb_microcdr/microRTPS_client.cpp.em

```diff
@@ -166,6 +196,49 @@ void *send(void *args)
 		}
 @[    end for]@
 
+@[    if send_topics_poll]@
+		{
+			px4_pollfd_struct_t fds[@(len(send_topics_poll))];
+@[        for idx, topic in enumerate(send_topics_poll)]@
+			fds[@(idx)].fd = subs_poll->@(topic)_sub;
+			fds[@(idx)].events = POLLIN;
+@[        end for]@
+
+			int pret = px4_poll(fds, @(len(send_topics_poll)), 20);
+
+			if (pret < 0) {
+				PX4_ERR("poll failed (%i)", pret);
+			} else if (pret != 0) {
+@[        for idx, topic in enumerate(send_topics_poll)]@
+				if (fds[@(idx)].revents & POLLIN) {
+					@(send_base_types_poll[idx])_s @(topic)_data;
+					orb_copy(ORB_ID(@(topic)), subs_poll->@(topic)_sub, &@(topic)_data);
+					serialize_@(send_base_types_poll[idx])(&writer, &@(topic)_data, &data_buffer[header_length], &length);
+
+					if (0 < (read = transport_node->write(static_cast<char>(@(msgs[0].index(topic) + 1)), data_buffer, length))) {
+						data->total_sent += read;
+						tx_last_sec_read += read;
+						++data->sent;
+					}
+				}
+@[        end for]@
+			} else {
+				if (hrt_absolute_time() - last_stats_update >= 1_s) {
+					data->sent_last_sec = tx_last_sec_read;
+					if (data->datarate > 0) {
+						bandwidth_mult = static_cast<float>(data->datarate) / static_cast<float>(tx_last_sec_read);
+						// Apply a low-pass filter to determine the new TX interval
+						tx_interval += 0.5f * (tx_interval / bandwidth_mult - tx_interval);
+						// Clamp the interval between 1 and 1000 ms
+						tx_interval = math::constrain(tx_interval, MIN_TX_INTERVAL_US, MAX_TX_INTERVAL_US);
+					}
+					tx_last_sec_read = 0;
+					last_stats_update = hrt_absolute_time();
+				}
+				px4_usleep(tx_interval);
```

Comment:
You can remove this here.
For the non-polled topics you can add a check if `tx_interval` time has passed since the last send, to avoid increasing their rates if there's a fast polled topic.

Reply:
I'm not sure I understand what you mean, sorry.  To avoid increasing the non-polled topics rates you mean? If this is what you mean, I wouldn't be increasing their rates since I'm calling update and only send this topic in case the update is true. Is that correct?

Reply:
> Do you mean to avoid increasing the non-polled topics rates?

Yes. The current behavior is: you poll on a topic and then run a full update loop. This means if your polled topic updates at a fast rate (and you don't restrict it via poll_interval), all others will be as well, without considering `tx_interval`. Does that make sense?

Reply:
Ok, so I may have a misunderstanding of the `poll` function. 

The way I see it for the **polled topics**:
1. Here, we poll all poll-topics inside the fds https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L207 
2. Then, we only send through RTPS the topics that have been published during the `poll` sleep. We distinguish among the the "updated" topics with this condition: https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L213 I am not sure whether with this condition we can check that. For the tests I did, it seemed to me that we can.

If this is true, then we would not need any sleep, right? Since we would only be publishing the messages that have been updated during the `poll` call. Thus every polled topic will be sent, at maximum, at the same rate it is being published in the uORB topic.

For the **non-polled topics**:
1. Before calling to `poll` we update all non-polled topics here: https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L169
2. If there's new data in the topic this function will return true and we will send it via RTPS. Otherwise, we continue.
3. If this is true, then we cannot send this topic at higher rates, since this function will return false if there are no new data.

Sorry for this long message. I wanted to be sure that I correctly understood the functionality behind each method.

Reply:
> If this is true, then we would not need any sleep, right?

Yes, correct.

> If this is true, then we cannot send this topic at higher rates, since this function will return false if there are no new data.

Yes correct as well.
I was refering to a case like this:
- `sensor_accel` polled, updated at 250Hz
- `vehicle_attitude` not polled. also updated at 250Hz. But assume we don't want to send it at 250Hz, then the above logic does not allow for it. Therefore I suggest to include a check for `tx_interval`, so that if you set it to very small (e.g. 1ms), you can get attitude at 250Hz as well, but you can increase the interval to reduce the sending rate of the attitude only, and still get sensor_accel at full rate.

Reply:
Ah ok! now I see the case that you meant. However, I'm not sure how to implement it. It seems to me that if we don't want to delay the whole loop we have to check `tx_interval` before this line:
https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L169

Would you mind giving me some hints on how to implement this?

Reply:
Exactly, a very simple way (with some inaccuracies) is be to add a check before https://github.com/PX4/PX4-Autopilot/blob/fb8725c335310b070f73ee45109cb423fe37b297/msg/templates/uorb_microcdr/microRTPS_client.cpp.em#L165
```
hrt_abstime last_update = 0;
while (!_should_exit_task) {
  hrt_abstime now = hrt_absolute_time();
  if (now - last_update > tx_interval) {
    last_update = now;
    ...
```

Reply:
Ah ok, the simple way indeed. Thanks for this. Done.